### PR TITLE
Add wall-clock cosine LR schedule (scheduler_interval: time)

### PR DESCRIPTION
**Pull request by farhanferoz**
> Stacked on #387 (base branch `fix-optimizer-config-defaults`). Review/merge #387 first; this diff shows only the wall-clock delta.

## Summary

Adds an opt-in `scheduler_interval: "time"` that anneals the cosine LR schedule over `trainer.max_time` (wall-clock) instead of a step/epoch horizon. The default epoch/step cosine is unchanged.

**Why:** today we fit the cosine to an *epoch count*, so on walltime-bounded jobs we run a `time-epochs` timing job to back-compute `max_epochs`. That's a two-step workflow, and if throughput drifts the schedule and the real job length diverge (LR bottoms out early, or never fully anneals before `max_time` stops the job). Wall-clock cosine removes the timing run, lands the LR floor exactly as the budget runs out, and self-corrects for throughput. This is the practical version of a well-known result: a cosine schedule is only optimal when its cycle length matches the training duration — too-long cycles undertrain (Hoffmann et al., *Chinchilla*, 2022, arXiv:2203.15556). Annealing over wall-clock makes that horizon the actual budget, automatically.

## Usage

Opt in with two settings — switch the cosine to the `time` interval and give the trainer a wall-clock budget:

```yaml
# optimizer config (adamw.yaml, or an experiment / CLI override)
scheduler: cosine
scheduler_interval: time
warmup: 0.05        # fraction of the budget; absolute counts are rejected in time mode
min_lr_ratio: 0.0   # optional LR floor

# trainer config
max_time: "00:12:00:00"   # required: the budget the cosine anneals over
```

Equivalently as CLI overrides:

```
autocast epd ... optimizer.scheduler_interval=time trainer.max_time="00:12:00:00" optimizer.warmup=0.05
```

No `time-epochs` timing run and no `max_epochs` needed — the LR reaches `min_lr_ratio` exactly as `max_time` is hit, and self-corrects if throughput drifts. For a smooth cosine across chained / resumed jobs, see the resume contract below.

## Mechanism

The `LambdaLR` lambda ignores the step counter and reads elapsed training time from Lightning's `Timer` each step: `progress = time_elapsed / (time_elapsed + time_remaining)`, cosine over `progress`. Lightning steps it per optimizer step (the `time` interval maps to `"step"` in `configure_optimizers`). `warmup` becomes a fraction of the budget. No new checkpoint state — the lambda just re-reads the Timer.

## Resume contract — would value a read on this, @sgreenbury

The schedule reads the **same `Timer` that enforces `max_time`**, whose elapsed offset is restored from checkpoints and zeroed by the timer-reset callback when `reset_resume_time_budget=true`. So resume behaviour falls out of how `max_time` already works:

- **Single-job run:** anneals over the job (the common case).
- **Chained run, smooth cosine:** set `max_time` to the *total* intended training time and full-state-resume (no reset) — the restored offset makes `time_elapsed` cumulative, so one cosine spans all jobs.
- **Chained run with `reset_resume_time_budget=true` each job:** each job gets a fresh `max_time` window, so the schedule restarts too — a **per-job sawtooth**. Consistent with the "fresh window" semantics of that flag, but it's a real choice.

I went with "the schedule follows `max_time`'s clock" rather than introducing a *separate* total-time budget decoupled from the per-job stop — one coherent knob, no new state, at the cost of the sawtooth above. **If you'd rather a chained run always be one smooth cosine regardless of the per-job reset, that needs a decoupled budget — happy to add it, but wanted to agree the contract first.**

## Scope / non-goals

- Just wall-clock cosine. Not WSD or Schedule-Free (separate experiments).
- Default behaviour unchanged; opt-in only.
- `cosine_with_restarts` + `time` and absolute (non-fractional) warmup in `time` mode are rejected with clear errors. Requires `trainer.max_time`.
- `time-epochs` is untouched; if this proves out we can decide whether to keep it as a fallback or deprecate.

## Testing

- 9 unit tests over a controllable `Timer`: progress→LR (0 / half / full / past-budget clamp), warmup-as-fraction, and three guard errors (missing budget, restarts, absolute warmup).
- A real `max_time` `Trainer` fit smoke confirming the Timer is found in `trainer.callbacks` and the `time→step` interval mapping is accepted.
- `uv run pytest tests/models/test_optimizer_mixin.py tests/scripts/test_training.py` → 66 passed. Ruff + pyright clean.

### CODE REVIEW

File: src/autocast/models/optimizer_mixin.py [hunk 1]
```diff
@@ -188,6 +279,16 @@ def _create_scheduler(
             raise ValueError(msg)
 
         if scheduler_name in {"cosine", "cosine_with_restarts"}:
+            if scheduler_interval == "time":
```
Comment (sgreenbury) [inline src/autocast/models/optimizer_mixin.py:282]: Perhaps we should raise an exception of `scheduler_interval=time` and `scheduler_name not in {"cosine", "cosine_with_restarts"}` (e.g. on L274?) so that non-supported pairs have no fallback?

**Review by sgreenbury — COMMENTED** [review]
Looks great! I think this fulfils the fixed time experimental design and solves related issues we've seen recently with variable throughput.

Just minor comments below for a later issue/PR
- There might be an issue when reset is set to `True`, the checkpoint is then loaded and then restores the scheduler's LR at that point for the the first step?
- Perhaps worth flagging somewhere that sharp changes might occur around validation  (this will typically be different length of time to training step) or if irregular (i.e. not every epoch). Also if the throughput changes during the run for some reason (e.g. the start we discussed, though could be elsewhere too). I think this is just something we should mention in the docs if using this scheduling approach for awareness.
